lib/commands/context.js

```diff
@@ -0,0 +1,472 @@
+                 'current window; assuming it is closed');
+    if (keyId !== null) {
+      logger.debug(`Debugger already selected page '${keyId}', ` +
+                   `confirming that choice.`);
+    } else {
+      logger.error('Do not have our current window anymore, and there ' +
+                   'are not any more to load! Doing nothing...');
+      return;
+    }
+    this.curContext = keyId;
+    // this.remote.pageIdKey = parseInt(keyId, 10);
+    newPage = keyId;
+  } else {
+    let dirty = () => {
+      var item = (arr) => {
+        return _.filter(arr, function (obj) {
+          return obj.id === this.curContext;
+        }, this)[0];
+      };
+
+      return !_.isEqual(item(this.contexts), item(pageArray));
+    };
+
+    // If a window gets navigated to an anchor it doesn't always fire a page callback event
+    // Let's check if we wound up in such a situation.
+    if (dirty()) {
+      this.remote.pageLoad();
+    }
+
+    logger.debug('New page listing is same as old, doing nothing');
+  }
+
+  if (newPage !== null) {
+    this.selectingNewPage = true;
+    await this.remote.selectPage(parseInt(newPage, 10));
+    this.selectingNewPage = false;
+    this.curContext = newPage;
+  }
+  this.windowHandleCache = _.map(pageArray, this.massagePage);
+};
+
+extensions.getLatestWebviewContextForTitle = async function (titleRegex) {
+  let contexts = await this.getContextsAndViews();
+  let matchingCtx;
+  // _(contexts).each((ctx) => {
+  for (let ctx of contexts) {
+    if (ctx.view && (ctx.view.title || '').match(titleRegex)) {
+      if (ctx.view.url === 'about:blank') {
+        // in the case of Xcode  < 5 (i.e., iOS SDK Version less than 7)
+        // and in the case of iOS 7.1 in a webview (not in Safari)
+        // we can have the url be `about:blank`
+        if (parseFloat(this.iOSSDKVersion) < 7 ||
+            (this.caps.platformVersion === '7.1' && this.caps.app && this.caps.app.toLowerCase() !== 'safari')) {
+          matchingCtx = ctx;
+        }
+      } else {
+        matchingCtx = ctx;
+      }
+    }
+  }
+  return matchingCtx ? matchingCtx.id : undefined;
+};
+
+// Right now we don't necessarily wait for webview
+// and frame to load, which leads to race conditions and flakiness
+// , let see if we can transition to something better
+extensions.useNewSafari = function () {
+  return parseFloat(this.iosSdkVersion) >= 8.1 &&
+         parseFloat(this.caps.platformVersion) >= 8.1 &&
+         !this.caps.udid &&
+         this.caps.safari;
+};
+
+extensions.navToInitialWebview = async function () {
+  let timeout = 0;
+  if (this.isRealDevice()) {
+    timeout = 3000;
+    logger.debug(`Waiting for ${timeout} ms before navigating to view.`);
+  }
+  await B.delay(timeout);
+  if (this.useNewSafari()) {
+    await this.typeAndNavToUrl();
+  } else if (parseInt(this.iosSdkVersion, 10) >= 7 && !this.caps.udid && this.caps.safari) {
```

Comment:
Then we should try to use them instead of `this.caps.udid` or `this.caps.safari`. I know is not done yet in a lot of places on master.